### PR TITLE
Testing bitflag expansion

### PR DESCRIPTION
**Pull request by jsantell**
Based off of @ncalexan's #214, relevant commit:

https://github.com/mozilla/mentat/pull/242/commits/a021735aefb65a6155fbc1628c2527e1cacd71fc

### CODE REVIEW

File: db/src/db.rs [hunk 1]
```diff
@@ -731,14 +732,19 @@ impl DB {
             .chain_err(|| "Could not update datoms: failed to retract datoms already present")?;
 
         // Insert datoms that were added and not already present.
-        let s = r#"
+        let s = format!(r#"
```
Comment (jsantell) [inline db/src/db.rs]: This is performance benchmarking, so someone should have an eye on this query other than me

File: db/src/db.rs [hunk 2]
```diff
@@ -37,6 +37,11 @@ pub const CURRENT_VERSION: i32 = 2;
 const TRUE: &'static bool = &true;
 const FALSE: &'static bool = &false;
 
+const INDEX_AVET     : u8 = 1 << 0;
```
Comment (ncalexan) [inline db/src/db.rs]: nit: for the final patch, comment on what these are and how the scheme works.

File: db/src/db.rs [hunk 3]
```diff
@@ -591,27 +587,35 @@ impl DB {
...
-                                   /* index_fulltext0 */ bool,
-                                   /* unique_value0 */ bool)>> = chunk.map(|&(e, a, ref typed_value, added)| {
+                                   /* flags */ u8)>> = chunk.map(|&(e, a, ref typed_value, added)| {
                 count += 1;
                 let attribute: &Attribute = self.schema.require_attribute_for_entid(&a)?;
 
                 // Now we can represent the typed value as an SQL value.
                 let (value, value_type_tag): (ToSqlOutput, i32) = typed_value.to_sql_value_pair();
 
+                let mut flags: u8 = 0;
```
Comment (ncalexan) [inline db/src/db.rs]: Consider implementing `Attribute.flags()` or similar.

File: db/src/db.rs [hunk 4]
```diff
@@ -620,18 +624,15 @@ impl DB {
                                   .chain(once(value_type_tag as &ToSql)
                                          .chain(once(&tx as &ToSql)
                                                 .chain(once(to_bool_ref(added) as &ToSql)
-                                                       .chain(once(to_bool_ref(index_avet) as &ToSql)
-                                                              .chain(once(to_bool_ref(index_vaet) as &ToSql)
-                                                                     .chain(once(to_bool_ref(index_fulltext) as &ToSql)
-                                                                            .chain(once(to_bool_ref(unique_value) as &ToSql))))))))))
+                                                       .chain(once(flags as &ToSql)))))))
```
Comment (ncalexan) [inline db/src/db.rs]: \o/

File: db/src/db.rs (hunk 1 above)
Comment (ncalexan) [inline db/src/db.rs]: Yeah, this is early days.  We're still deep in the "make it correct" phase, but I was hoping to get some idea of how much this saves.  We'll be better able to benchmark as the transactor fleshes out.

File: db/src/db.rs [hunk 5]
```diff
@@ -518,10 +523,7 @@ impl DB {
                value_type_tag0 SMALLINT NOT NULL,
```
Comment (ncalexan) [inline db/src/db.rs]: These tables definitely deserve a comment about the flag compression.  And flags should still be a `TINYINT`, although this doesn't make much difference in SQLite, IIUC.

**Comment by jsantell** [comment]
@ncalexan rebased on that big PR. For implementing `Attribute.flags()`, to clarify, are you talking about:

* Moving BitFlags type to core/src/lib and importing that into db/src/db
* Provide an `Attribute.flags()` function that returns a bitfield containing all of the flags (index_vaet, etc) that we want in the bitfield?

**Comment by ncalexan** [comment]
> Provide an Attribute.flags() function that returns a bitfield containing all of the flags (index_vaet, etc) that we want in the bitfield?

@jsantell, I was thinking of this.

**Comment by jsantell** [comment]
@ncalexan  Ok, made the type changes and `flags()` helper -- good for review

File: core/src/lib.rs [hunk 6]
```diff
@@ -81,6 +81,17 @@ fn test_typed_value() {
...
 }
 
+/// Bit flags used in `flags0` column in temporary tables created during search,
+/// such as the `search_results`, `inexact_searches` and `exact_searches` tables.
+/// When moving to a more concrete table, such as `datoms`, they are expanded out
+/// via these flags and put into their own column rather than a bit field.
+pub enum AttributeBitFlags {
+    IndexAVET     = 1 << 0,
+    IndexVAET     = 1 << 1,
+    IndexFullText = 1 << 2,
```
Comment (ncalexan) [inline core/src/lib.rs]: nit: I think we're considering "fulltext" one word, but I'm not militant about this.

File: db/src/db.rs [hunk 7]
```diff
@@ -56,6 +56,8 @@ const TRUE: &'static bool = &true;
 const FALSE: &'static bool = &false;
 
 /// Turn an owned bool into a static reference to a bool.
```
Comment (ncalexan) [inline db/src/db.rs]: Spurious change?

File: db/src/db.rs [hunk 8]
```diff
@@ -537,31 +531,27 @@ impl DB {
...
 
+                let flags = attribute.flags();
+
                 Ok((e, a, value, value_type_tag,
                     added,
-                    attribute.index,
-                    attribute.value_type == ValueType::Ref,
-                    attribute.fulltext,
-                    attribute.unique_value))
+                    flags))
```
Comment (ncalexan) [inline db/src/db.rs]: Maybe just inline this call?  Your choice, your style.

**Comment by ncalexan** [comment]
Also, check the tests of all the crates -- Travis is angry :)

File: core/src/lib.rs (hunk 6 above)
Comment (jsantell) [inline core/src/lib.rs]: Good call

Comment (jsantell) [inline core/src/lib.rs]: And looks like this is the best way to do this according to #rust

**Pull request merged**